Fri05

Noordfrees opened the pull request:
The lengthiest and probably most difficult mission ever written so far, complete at last. This scenario initially gives you a few advantages … and every imaginable disadvantage. Luckily you have smart advisors who will help you, but this mission still requires attentive strategic planning and some non-standard tactics on your part or you will either lose rapidly or your economy will get stuck beyond repair. 

The main part of the scenario however is about sustaining an economy for an extended period of time. This is a gimmick that has not been used before and which should not be used again, but having 1 scenario like this seems just right.

There's a codechange for eyecandy here. This feature is not available in the editor or in other maps generally because it can cause masses of bugs when used carelessly; every node on this map has been hand-crafted individually to ensure it will work fine here.

Storyline spoiler and strategic tips will be added presently at https://www.widelands.org/wiki/FrisiansStory/.

Noordfrees committed:
Support custom terrains in scenarios

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Allow steeper slopes, if the map designer knows what he's doing

Noordfrees committed:
Start to work on fri05

Noordfrees committed:
initial scripting

Noordfrees committed:
Implement trading and fishing

Noordfrees committed:
Fix rendering glitches, mostly

Noordfrees committed:
Finish base implementation

Noordfrees committed:
two more tweaks

Noordfrees committed:
Add Lua mutex to fix a (possibly unrelated) crash

Noordfrees committed:
Misc tweaks

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge ../widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Fix optional parentheses

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Fixes an updates

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Update with more objectives and dialog

Noordfrees committed:
Fixes

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Minor difficulty tweak

Noordfrees committed:
Remove comment

Noordfrees committed:
3 files were automatically formatted.

Noordfrees committed:
Fix clang-tidy hit

hessenfarmer reviewed (commented):
just a first bunch of findings after my first round of testplaying.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Address reviee

Noordfrees committed:
Fix soldier takeover bug

hessenfarmer commented:
another thing I found not really fitting is the background picture containing a concrete dike. A picture from the Helgoland rocks without modern artificial elements would look more fitting to the story. but this is not that important

Noordfrees commented:
Unfortunately such an image might be hard to come by, the whole island looks like a huge military harbour overcrowded with tourists. This photo is the least occluded one I have that shows anything recognizable as typical for Helgoland (the rock spur is its landmark the Lange Anna), all photos with no artificial objects are just close-up shots of seabirds nesting on red-and-white rocks.
![DSC07675](https://user-images.githubusercontent.com/48293446/208666585-65df5e9f-9f85-4233-9096-305dfc9f0173.JPG)
![DSC07690](https://user-images.githubusercontent.com/48293446/208667400-7af2f93c-6a7d-4c42-a2d6-e83441286920.JPG)

hessenfarmer commented:
ok some further test playing showed especially one annoying behaviour. Due to the eye candy feature it is impossible to use some of the building plots at the northern border of the "uplands". loading the map in the editor (figured out how to do that ;-)) shows really weird grid lines there. for example it is impossible to get a click on the big building plot which is drawn as being on 55,83.   
for me this is a showstopper as a normal player can't know why and would most likely report  a bug.

hessenfarmer commented:
regrding the image you may be right, I could not find any image fitting to our license on th internet. However on helgoland.de there are some pictures without external reference, so at least it might be worth a try to contact the local government of the island and ask for a fitting picture as the scenario is to some extent some free advertising for the island. I think as chieftain of widelands you could try to write them a mail. 
Else I could live with the current picture and maybe we might use gimp to just remove the concrete or to get it a texture of rough stones making it look like an medieval breakwater.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Fix clicking on obscuring fields

Noordfrees committed:
Change background image

Noordfrees commented:
Okay I found one image that doesn't have any manmade structures and still at least shows the red-and-white rocks and the seabirds and the high seas in the background. How about that?

hessenfarmer commented:
> Okay I found one image that doesn't have any manmade structures and still at least shows the red-and-white rocks and the seabirds and the high seas in the background. How about that?

I am ok with it if you are content with it. However the "Lange Anna" would be nicer of course as being a gret signature landmark. So your decision which you deem best fitting to location and story.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Mark fighting objective as done after enemy defeated

hessenfarmer commented:
I found a severe bug in the map cursor code. to reproduce either in the editor or in a game try to move the cursor over the zero lines in the map.

Noordfrees committed:
Fix map pixel coords wraparound bug

Noordfrees committed:
Change portraits

Noordfrees commented:
Fixed the bug and changed the image borders. Ketelsen now uses Hauke's image. Instead of adding the name to the heading how about adding a plaque below each portrait with the name on it? That should be done in a separate branch though as it should be changed consistently across all frisian scenarios.

I tried to implement the revealing suggestion but I couldn't get it to work properly. Scripted hiding keeps overriding regular vision and never does what I'd expect it should :( 

Heligoland is an alternative spelling to Helgoland, both are [valid](https://dict.leo.org/englisch-deutsch/helgoland) in English. I'd prefer to stick to the spelling that is identical to the German spelling. We also have Helgoland as a Frisian shipname that uses this spelling.

hessenfarmer commented:
+1 for the plaque.  
for the revelation code I am already working on it. so expect a branch soon

hessenfarmer committed:
alternative field revelation at the beginning

hessenfarmer commented:
Ok PR with the suggestion for revealing the fields is up.    
this contains a new camera move as well.   
now I will work on the other suggestuions I made

Noordfrees committed:
Some code cleanup and fixes

hessenfarmer committed:
reordered texts and made some dependent on game

hessenfarmer committed:
fixed syntax error

hessenfarmer commented:
ok I made a PR with some textual suggestions. see if you like it and feel free to alter them.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Merge branch 'fri05_some_suggestions' of github.com:widelands/widelands into fri05

Noordfrees committed:
Some small rephrasings and fixes

Noordfrees commented:
Thanks :)

vworldat commented:
This looks awesome so far, thanks for your great work!

Playing the latest f623fa800d build I caught a lua script error that appeared as soon as the first wood cutters and foresters had been built:

Edit: figured out the log shows the same message, removed the screen shot.

```log
[00:16:38.537 game] ERROR: [../src/scripting/lua_errors.cc:21] [string "scripting/mission_thread.lua"]:79: attempt to compare number with boolean
```

Noordfrees committed:
Fix operator precedence bug

Noordfrees commented:
Thanks :) Bug should be fixed

vworldat commented:
Another one:
```log
[01:15:05.661 game] ERROR: [../src/scripting/lua_errors.cc:21] [string "scripting/mission_thread.lua"]:55: attempt to perform arithmetic on a nil value (global 'tradecount')
```

Noordfrees committed:
Fix variable name typo

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

hessenfarmer reviewed (commented):
some last remarks inline.
Furthermore I would love to increase the height of message boxes with objectives to see there is a new objective.    
Last one is that we scroll to certain points for displaying a message box but then we obscure the center with the box, so it might be good to move the box to the left or right or even in one of the corners.

Noordfrees committed:
Implement suggestions

Noordfrees committed:
Make objective message boxes taller and move some of them to the screen side

hessenfarmer commented:
I checked the moved textboxes now and changing the position from the middle to topleft and back feels uncomfortable. may we have all of them topleft (although for me toprigth would feel better but that is a matter of taste).  
if you prefer to keep the mix the training nd diking box should be moved topleft at least as we move to a special point to show them.
rest is fine now from my side. 
off topic: I am not happy with the AI perfoirmance in this scenario as it makes some mistakes that can be used to win it more easy. ( most of all it still keeps dismantling milsites without sense)

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Move training and diking message boxes topleft too

hessenfarmer reviewed (approved):
LGTM now.
only thing is it might be good to have a proofreadiong of the texts. But I have seen them so often now I can't find anything. 
shall we ask @tothxa or @frankystone

tothxa reviewed (commented):
Just a few grammatical suggestions.

Feel free to ignore my comments on possible anachronisms. I can see that it would be hard to clean them up.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into fri05

Noordfrees committed:
Rephrasings

tothxa reviewed (approved):
Thank you.

Noordfrees commented:
Thanks for all the playtesting and the reviewing :)
